pkg/metrics/builder/builder_test.go

```diff
@@ -0,0 +1,221 @@
+				assert.NoError(t, err, "expected successful builder creation")
+			} else {
+				assert.Error(t, err, "expected builder creation failure but builder was created")
+				assert.EqualError(t, err, tc.expectedErrorMessage)
+			}
+		})
+	}
+}
+
+func TestMetricBuilderCollector(t *testing.T) {
+	cases := []struct {
+		name                  string
+		expectedCollectorType string
+		opts                  MetricOpts
+		labelKeyValues        map[string]string
+		value                 float64
+		expectedErrorMessage  string
+	}{
+		{
+			name:                  "Test histogram creation",
+			expectedCollectorType: "histogram",
+			opts: MetricOpts{
+				Labels:     []string{"test1", "test2"},
+				Desc:       "test histogram metrics",
+				Name:       "test_histogram",
+				Buckets:    []float64{10, 20, 30},
+				MetricType: Histogram,
+			},
+			labelKeyValues:       map[string]string{"test1": "test1", "test2": "test2"},
+			value:                0,
+			expectedErrorMessage: "",
+		},
+		{
+			name:                  "Test counter creation",
+			expectedCollectorType: "counter",
+			opts: MetricOpts{
+				Labels:     []string{"test1", "test2"},
+				Desc:       "test modification metric",
+				Name:       "test_modification",
+				MetricType: Counter,
+			},
+			labelKeyValues:       map[string]string{"test1": "test1", "test2": "test2"},
+			expectedErrorMessage: "",
+		},
+		{
+			name:                  "Test empty label values",
+			expectedCollectorType: "counter",
+			opts: MetricOpts{
+				Labels:     []string{"test1", "test2"},
+				Desc:       "test modification metric",
+				Name:       "test_modification",
+				MetricType: Counter,
+			},
+			labelKeyValues:       nil,
+			value:                0,
+			expectedErrorMessage: "",
+		},
+		{
+			name:                  "Test invalid metricType",
+			expectedCollectorType: "Linear",
+			opts: MetricOpts{
+				Labels:     []string{"test1", "test2"},
+				Desc:       "test metric",
+				Name:       "test",
+				MetricType: "Linear",
+			},
+			labelKeyValues:       map[string]string{"test1": "test1", "test2": "test2"},
+			value:                0,
+			expectedErrorMessage: `invalid metric builder type "Linear". cannot create collector`,
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			metricBuilder, err := NewMetricBuilder(tc.opts, tc.value, tc.labelKeyValues)
+			collector, err := metricBuilder.PromCollector()
+			if tc.expectedErrorMessage == "" {
+				assert.NoError(t, err, "expected successful collector creation")
+				assert.EqualValues(t, tc.expectedCollectorType, reflect.TypeOf(collector).Elem().Name())
+				expectedString := getDescStruct(tc.opts, tc.labelKeyValues)
+				assert.EqualValues(t, expectedString, getCollectorDescription(collector))
+			} else {
+				assert.Error(t, err, "expected collector creation failure but collector was created")
+				assert.EqualError(t, err, tc.expectedErrorMessage)
+			}
+		})
+	}
+}
+func TestNewMetricBuilder(t *testing.T) {
```

Comment:
Can we add a comment:

TestNewMetricBuilder tests the Metric Builder initializer.

Although I see now that this tests more than the iniitializer. I'm not sure if these are the most useful test cases. They are fine, but they don't test realistic errors/failure modes so I'm not sure if these add much value.

Reply:
Hmm.. do you have any kind of tests in mind? I thought that since this is just a struct with the metric type, there can only be some kind of validation tests and the significant tests can be done only when the API system is in place.